core/client/hammer/hammer.go

```diff
@@ -76,125 +97,199 @@ func New(ctx context.Context, dial DialFunc, callOptions CallOptions,
+		for i := range workers {
+			handlers = append(handlers, workers[i].readOp)
+		}
+		executeRequests(ctx, args, handlers)
+		fmt.Print("\n")
+	}
+
+	if ok := c.TestTypes["audit"]; ok {
+		// History
+		log.Printf("User Audit History")
+		args := genArgs(ctx, c.HistoryQPS, c.HistoryPageSize, c.HistoryCount, c.Duration)
+		handlers := make([]ReqHandler, 0, len(workers))
+		for i := range workers {
+			handlers = append(handlers, workers[i].historyOp)
+		}
+		executeRequests(ctx, args, handlers)
+		fmt.Print("\n")
+	}
 
-	return outputs
+	return nil //pool.Stop()
 }
 
-// gen produces a channel of integers from 0 up to count-1 and then closes.
-func gen(ctx context.Context, count int) <-chan int {
-	jobs := make(chan int)
+func genArgs(ctx context.Context, qps, batch, count int, duration time.Duration) <-chan opArg {
+	inflightReqs := make(chan opArg, qps)
 	go func() {
-		defer close(jobs)
+		cctx, cancel := context.WithTimeout(ctx, duration)
+		defer cancel()
+		defer close(inflightReqs)
+
+		rateLimiter := rate.NewLimiter(rate.Limit(qps), qps)
 		for i := 0; i < count; i++ {
-			select {
-			case <-ctx.Done():
+			userIDs := make([]string, 0, batch)
+			for j := 0; j < batch; j++ {
+				userIDs = append(userIDs, fmt.Sprintf("user_%v", i*batch+j))
+			}
+			if err := rateLimiter.WaitN(cctx, batch); err != nil {
+				log.Printf("Duration %v elapsed, stopping request generation", duration)
 				return
-			case jobs <- i:
+			}
+			inflightReqs <- opArg{
+				UserIDs:  userIDs,
+				PageSize: batch,
 			}
 		}
 	}()
-	return jobs
+	return inflightReqs
 }
 
-// printStats prints stats repeatedly every period until outputs closes.
-func (h *Hammer) printStats(period time.Duration, outputs <-chan error) {
-	startTime := time.Now()
-	var totalRequests float64
-
-	// Print stats asyncronously until refresh closes.
-	refresh := time.NewTicker(period)
-	stop := make(chan struct{})
-	var wg sync.WaitGroup
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		for {
-			select {
-			case <-refresh.C:
-				rate := totalRequests / time.Since(startTime).Seconds()
-				log.Printf("QPS: %v, \tRequests: %v, \tWorkers: %v", rate, totalRequests, h.workers)
-			case <-stop:
-				rate := totalRequests / time.Since(startTime).Seconds()
-				log.Printf("QPS: %v, \tRequests: %v, \tWorkers: %v", rate, totalRequests, h.workers)
-				refresh.Stop()
-				return
-			}
-		}
-	}()
+type worker struct {
+	*Hammer
+	client *client.Client
+	bar    *cpb.ProgressBar
+}
 
-	// Collect outputs until output closes.
-	for err := range outputs {
-		totalRequests++
+func (h *Hammer) newWorkers(n int) ([]worker, error) {
+	workers := make([]worker, 0, n)
+	for i := 0; i < n; i++ {
+		// Give each worker its own client.
+		client, err := client.NewFromConfig(h.ktCli, h.domain)
 		if err != nil {
-			log.Printf("writeOp(): %v", err)
+			return nil, err
 		}
+
+		workers = append(workers, worker{
+			Hammer: h,
+			client: client,
+			bar:    cpb.New(100).Prefix(fmt.Sprintf("%v ", i)),
```

Comment:
Do I understand correctly that in case of default 1000 workers you will have 1000 progress bars in the command line?

Have you considered doing just one? Or one for each type of hammering?

Reply:
Yes, that's correct. However, I'm removing the progress bar since 1000 progress bars aren't that user friendly ;-)